### PR TITLE
Super agi api

### PR DESCRIPTION
**Pull request by Maverick-F35**
### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update

### Checklist
- [x] My pull request is atomic and focuses on a single change.
- [ ] I have read the contributing guide and my code conforms to the guidelines.
- [ ] I have documented my changes clearly and comprehensively.
- [ ] I have added the required tests.

### CODE REVIEW

**Review by Fluder-Paradyne — CHANGES_REQUESTED** [review]

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('api_key',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('org_id', sa.Integer(), nullable=True),
+    sa.Column('key_name', sa.String(), nullable=True),
+    sa.Column('key', sa.String(), nullable=True),
+    sa.Column('created_at', sa.DateTime(), nullable=True),
+    sa.Column('updated_at', sa.DateTime(), nullable=True),
+    sa.Column('revoked',sa.Boolean(),nullable=True,default=False),
```
Comment (TransformerOptimus) [inline migrations/versions/446884dcae58_add_api_key_and_web_hook.py:28]: rename it to is_expired

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    )
+    op.create_table('web_hooks',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('name', sa.String(), nullable=True),
+    sa.Column('org_id', sa.Integer(), nullable=True),
+    sa.Column('url', sa.String(), nullable=True),
+    sa.Column('headers', sa.JSON(), nullable=True),
+    sa.Column('created_at', sa.DateTime(), nullable=True),
+    sa.Column('updated_at', sa.DateTime(), nullable=True),
+    sa.Column('isDeleted',sa.Boolean(),nullable=True),
```
Comment (TransformerOptimus) [inline migrations/versions/446884dcae58_add_api_key_and_web_hook.py:39]: should be is_deleted.

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    )
+
+    #add index *********************
+    # ### end Alembic commands ###
+
+
+def downgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    
+    op.drop_table('web_hooks')
```
Comment (TransformerOptimus) [inline migrations/versions/446884dcae58_add_api_key_and_web_hook.py:61]: webhooks is a single word. same for webhook_events

File: superagi/models/resource.py [hunk 4]
```diff
@@ -58,6 +58,11 @@ def validate_resource_type(storage_type):
 
         if storage_type not in valid_types:
             raise InvalidResourceType("Invalid resource type")
-
+    
+    @classmethod
+    def get_all_resources_from_run_ids(cls,session,run_ids_arr):
```
Comment (Fluder-Paradyne) [inline superagi/models/resource.py:63]: name change to execution_ids

File: superagi/models/agent_execution.py [hunk 5]
```diff
@@ -164,4 +164,61 @@ def assign_next_step_id(cls, session, agent_execution_id: int, next_step_id: int
...
+                "goal": agent_execution.goal,
+            }
+        
+        if agent_execution.instruction is not None:
+            agent_execution_configs["instructions"]=agent_execution.instruction,
+        return agent_execution_configs
+    
+    @classmethod
+    def get_all_executions_with_status_and_agent_id(cls,session,agent_id,ExecutionStateChangeConfigIn,current_status):
+        db_execution_arr=[]
```
Comment (Fluder-Paradyne) [inline superagi/models/agent_execution.py:188]: ["CREATED", "RUNNING", "PAUSED", "COMPLETED", "TERMINATED"]

File: superagi/helper/webhook_manager.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        agent_id=AgentExecution.get_agent_execution_from_id(self.session,agent_execution_id).agent_id
+        agent=Agent.get_agent_from_id(self.session,agent_id)
+        org=agent.get_agent_organisation(self.session)
+        org_id=org.id
+        org_webhooks=self.session.query(WebHooks).filter(org_id==org_id).all()
+
+        for webhook_obj in org_webhooks:
+            webhook_obj_body={"agent_id":agent_id,"org_id":org_id,"event":f"{old_val} to {val}"}
+            error=None
+            r=None
```
Comment (Fluder-Paradyne) [inline superagi/helper/webhook_manager.py:24]: r to request

File: superagi/helper/s3_helper.py [hunk 7]
```diff
@@ -111,4 +111,28 @@ def upload_file_content(self, content, file_path):
         try:
             self.s3.put_object(Bucket=self.bucket_name, Key=file_path, Body=content)
         except:
-            raise HTTPException(status_code=500, detail="AWS credentials not found. Check your configuration.")
\ No newline at end of file
+            raise HTTPException(status_code=500, detail="AWS credentials not found. Check your configuration.")
+        
+    @classmethod
+    def get_download_url_of_resources(cls,db_resources_arr):
+        s3 = boto3.client(
```
Comment (Fluder-Paradyne) [inline superagi/helper/s3_helper.py:118]: remove client creation and replace with self.s3

File: superagi/controllers/webhook.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+from superagi.helper.auth import check_auth,validate_api_key
+from superagi.models.agent import Agent
+from superagi.models.agent_execution_config import AgentExecutionConfiguration
+from superagi.models.agent_config import AgentConfiguration
+from superagi.models.agent_schedule import AgentSchedule
+from superagi.models.agent_template import AgentTemplate
+from superagi.models.project import Project
+from superagi.models.agent_execution import AgentExecution
+from superagi.models.tool import Tool
+from superagi.models.web_hooks import WebHooks
```
Comment (Fluder-Paradyne) [inline superagi/controllers/webhook.py:24]: remove unused imports

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py (hunk 1 above)
Comment (Maverick-F35) [reply migrations/versions/446884dcae58_add_api_key_and_web_hook.py]: renamed

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py (hunk 2 above)
Comment (Maverick-F35) [reply migrations/versions/446884dcae58_add_api_key_and_web_hook.py]: renamed

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py (hunk 3 above)
Comment (Maverick-F35) [reply migrations/versions/446884dcae58_add_api_key_and_web_hook.py]: changed

File: superagi/models/resource.py (hunk 4 above)
Comment (Maverick-F35) [reply superagi/models/resource.py]: done

File: superagi/helper/webhook_manager.py (hunk 6 above)
Comment (Maverick-F35) [reply superagi/helper/webhook_manager.py]: renamed

File: superagi/helper/s3_helper.py (hunk 7 above)
Comment (Maverick-F35) [reply superagi/helper/s3_helper.py:117]: done

File: superagi/controllers/webhook.py (hunk 8 above)
Comment (Maverick-F35) [reply superagi/controllers/webhook.py:15]: removed unused imports

File: superagi/models/agent_schedule.py [hunk 9]
```diff
@@ -45,4 +45,27 @@ def __repr__(self):
...
+            status="SCHEDULED"
+        )
+
+        agent_schedule.agent_id = db_agent.id
+        session.add(agent_schedule)
+        session.commit()
+        return agent_schedule
+    
+    @classmethod
+    def get_schedule_from_agent_id(self,session,agent_id):
```
Comment (TransformerOptimus) [inline superagi/models/agent_schedule.py:69]: change it to find_by_agent_id

File: superagi/models/agent_schedule.py [hunk 10]
```diff
@@ -45,4 +45,27 @@ def __repr__(self):
                f"expiry_date={self.expiry_date}, " \
                f"expiry_runs={self.expiry_runs}), " \
                f"current_runs={self.expiry_runs}), " \
-               f"status={self.status}), " 
\ No newline at end of file
+               f"status={self.status}), " 
+    
+    @classmethod
```
Comment (TransformerOptimus) [inline superagi/models/agent_schedule.py:50]: why u are doing save in get_schedule

File: superagi/models/agent_schedule.py (hunk 9 above)
Comment (TransformerOptimus) [reply superagi/models/agent_schedule.py:69]: cls not self.

File: superagi/models/agent_execution.py [hunk 11]
```diff
@@ -164,4 +164,61 @@ def assign_next_step_id(cls, session, agent_execution_id: int, next_step_id: int
...
+    def validate_run_ids(cls,session,run_ids_arr,organisation_id):
+        from superagi.models.agent import Agent
+        from superagi.models.project import Project
+        for run_id in run_ids_arr:
+            db_execution=session.query(AgentExecution).filter(AgentExecution.id==run_id).first()
+
+            if db_execution is None:
+                raise Exception(f"Run ID {run_id} not found")
+            
+            agent_id=db_execution.agent_id
```
Comment (TransformerOptimus) [inline superagi/models/agent_execution.py:218]: remove this extra variables.

File: superagi/models/agent_execution.py [hunk 12]
```diff
@@ -164,4 +164,61 @@ def assign_next_step_id(cls, session, agent_execution_id: int, next_step_id: int
...
+            db_execution_arr=session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id,AgentExecution.status==current_status,AgentExecution.id.in_(ExecutionStateChangeConfigIn.run_ids)).all()
+        else:
+            db_execution_arr=session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id,AgentExecution.status==current_status).all()
+        return db_execution_arr
+    
+    @classmethod
+    def get_all_executions_with_filter_config_and_agent_id(cls,session,agent_id,filter_config):
+        db_execution_arr=[]
+        if filter_config.run_ids is not None and filter_config.run_status_filter is not None and filter_config.run_status_filter in ["CREATED", "RUNNING", "PAUSED", "COMPLETED", "TERMINATED"]:
+            db_execution_arr=session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id,AgentExecution.id.in_(filter_config.run_ids),AgentExecution.status==filter_config.run_status_filter).all()
```
Comment (TransformerOptimus) [inline superagi/models/agent_execution.py:199]: query = session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id)

if condition1:
   query = query.filter(AgentExecution.id.in_(filter_config.run_ids))
   
if condition2:
   query = query.filter(AgentExecution.id.in_(filter_config.run_ids))
 
rename it find_executions_by_filter_config

File: superagi/models/agent_execution.py [hunk 13]
```diff
@@ -164,4 +164,61 @@ def assign_next_step_id(cls, session, agent_execution_id: int, next_step_id: int
...
\ No newline at end of file
+        session.commit()
+
+    @classmethod
+    def get_execution_from_agent_id_with_status(cls,session,agent_id,status_filter):
+        db_agent_execution = session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id,AgentExecution.status==status_filter).first()
+        return db_agent_execution
+    
+    @classmethod
+    def get_updated_execution_config_obj(cls,agent_execution):
```
Comment (TransformerOptimus) [inline superagi/models/agent_execution.py:175]: move it to controller

File: superagi/helper/webhook_manager.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+from superagi.models.web_hooks import WebHooks
+from superagi.models.web_hook_events import WebHookEvents
+import requests
+import json
+from superagi.lib.logger import logger
+class WebHookManager:
+    def __init__(self,session):
+        self.session=session
+
+    def agentStatusChangeCallback(self,agent_execution_id, val,old_val):
```
Comment (TransformerOptimus) [inline superagi/helper/webhook_manager.py:12]: snake case

File: superagi/helper/webhook_manager.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+        if val=="CREATED" or agent_execution_id is None:
+            return
+        agent_id=AgentExecution.get_agent_execution_from_id(self.session,agent_execution_id).agent_id
+        agent=Agent.get_agent_from_id(self.session,agent_id)
+        org=agent.get_agent_organisation(self.session)
+        org_id=org.id
```
Comment (TransformerOptimus) [inline superagi/helper/webhook_manager.py:18]: remove org_id tmp

File: superagi/helper/webhook_manager.py [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+            try:
+                request = requests.post(webhook_obj.url.strip(), data=json.dumps(webhook_obj_body), headers=webhook_obj.headers)
+            except Exception as e:
+                logger.error(f"Exception occured in webhooks {e}")
+                error=str(e)
+            if request is not None and request.status_code not in [200,201] and error is None:
+                error=request.text
+            if error is not None:
+                status='Error'
+            web_hook_event=WebHookEvents(agent_id=agent_id,run_id=agent_execution_id,event=f"{old_val} to {val}",status=status,errors=error)
```
Comment (TransformerOptimus) [inline superagi/helper/webhook_manager.py:35]: webhook_event

File: superagi/helper/webhook_manager.py [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+            self.session.add(web_hook_event)
+            self.session.commit()
+        return
```
Comment (TransformerOptimus) [inline superagi/helper/webhook_manager.py:38]: not required

File: superagi/models/resource.py [hunk 18]
```diff
@@ -58,6 +58,11 @@ def validate_resource_type(storage_type):
 
         if storage_type not in valid_types:
             raise InvalidResourceType("Invalid resource type")
-
+    
+    @classmethod
+    def get_all_resources_from_run_ids(cls,session,execution_ids):
```
Comment (TransformerOptimus) [inline superagi/models/resource.py:63]: find_resources_by_run_ids

rename execution_ids param to run_ids

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('api_key',
+    sa.Column('id', sa.Integer(), nullable=False),
+    sa.Column('org_id', sa.Integer(), nullable=True),
+    sa.Column('key_name', sa.String(), nullable=True),
```
Comment (TransformerOptimus) [inline migrations/versions/446884dcae58_add_api_key_and_web_hook.py:24]: rename key_name to name

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+# revision identifiers, used by Alembic.
+revision = '446884dcae58'
+down_revision = '2fbd6472112c'
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('api_key',
```
Comment (TransformerOptimus) [inline migrations/versions/446884dcae58_add_api_key_and_web_hook.py:21]: should be api_keys

File: superagi/models/project.py [hunk 21]
```diff
@@ -55,3 +55,13 @@ def find_or_create_default_project(cls, session, organisation_id):
         else:
             default_project = project
         return default_project
+
+    @classmethod
+    def get_project_from_org_id(cls,session,org_id):
```
Comment (TransformerOptimus) [inline superagi/models/project.py:60]: find_by_org_id

File: superagi/models/project.py [hunk 22]
```diff
@@ -55,3 +55,13 @@ def find_or_create_default_project(cls, session, organisation_id):
...
+        project=session.query(Project).filter(Project.organisation_id==org_id).first()
+        return project
+    
+    @classmethod
+    def get_project_from_id(cls,session,project_id):
```
Comment (TransformerOptimus) [inline superagi/models/project.py:65]: find_by_id

File: superagi/models/toolkit.py [hunk 23]
```diff
@@ -138,3 +138,29 @@ def fetch_tool_ids_from_toolkit(cls, session, toolkit_ids):
...
+            toolkits_arr.add(toolkit.id)
+            if tool_obj.get("tools"):
+                for tool_name_str in tool_obj["tools"]:
+                    tool_db_obj=session.query(Tool).filter(Tool.name==tool_name_str.strip()).first()
+                    if tool_db_obj is None:
+                            raise Exception("One or more of the Tool(s)/Toolkit(s) does not exist.")
+
+                    tools_arr.add(tool_db_obj.id)
+            else:
+                tools=session.query(Tool).filter(Tool.toolkit_id==toolkit.id).all()
```
Comment (TransformerOptimus) [inline superagi/models/toolkit.py:160]: use get_toolkit_tools

File: superagi/models/toolkit.py [hunk 24]
```diff
@@ -138,3 +138,29 @@ def fetch_tool_ids_from_toolkit(cls, session, toolkit_ids):
...
+        toolkits_arr=set()
+        tools_arr=set()
+
+        for tool_obj in agent_config_tools_arr:
+            toolkit=session.query(Toolkit).filter(Toolkit.name==tool_obj["name"].strip()).first()
+            if toolkit is None:
+                raise Exception("One or more of the Tool(s)/Toolkit(s) does not exist.")
+            toolkits_arr.add(toolkit.id)
+            if tool_obj.get("tools"):
+                for tool_name_str in tool_obj["tools"]:
```
Comment (TransformerOptimus) [inline superagi/models/toolkit.py:153]: use convert_tool_names_to_ids

File: superagi/models/web_hook_events.py [hunk 25]
```diff
@@ -0,0 +1,25 @@
+from sqlalchemy import Column, Integer, Text, String, Boolean, ForeignKey
+from sqlalchemy.orm import relationship
+from superagi.models.base_model import DBBaseModel
+from superagi.models.agent_execution import AgentExecution
+
+
+class WebHookEvents(DBBaseModel):
```
Comment (TransformerOptimus) [inline superagi/models/web_hook_events.py:7]: WebhookEvents

File: superagi/models/agent_schedule.py (hunk 9 above)
Comment (Maverick-F35) [reply superagi/models/agent_schedule.py:69]: renamed and replaced self with cls

File: superagi/models/agent_schedule.py (hunk 10 above)
Comment (Maverick-F35) [reply superagi/models/agent_schedule.py:50]: renamed

File: superagi/models/agent_execution.py [hunk 26]
```diff
@@ -164,4 +164,61 @@ def assign_next_step_id(cls, session, agent_execution_id: int, next_step_id: int
...
+            db_execution_arr=session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id,AgentExecution.status==filter_config.run_status_filter).all()
+        else:
+            db_execution_arr=session.query(AgentExecution).filter(AgentExecution.agent_id==agent_id)
+        return db_execution_arr
+    
+    @classmethod
+    def validate_run_ids(cls,session,run_ids_arr,organisation_id):
+        from superagi.models.agent import Agent
+        from superagi.models.project import Project
+        for run_id in run_ids_arr:
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/models/agent_execution.py:212]: removed the loop and optimized the queries

File: superagi/models/agent_execution.py (hunk 12 above)
Comment (Maverick-F35) [reply superagi/models/agent_execution.py:199]: optimized and renamed

File: superagi/models/agent_execution.py (hunk 13 above)
Comment (Maverick-F35) [reply superagi/models/agent_execution.py:175]: removed from model and moved to controller

File: superagi/helper/webhook_manager.py (hunk 15 above)
Comment (Maverick-F35) [reply superagi/helper/webhook_manager.py:18]: removed

File: superagi/helper/webhook_manager.py (hunk 16 above)
Comment (Maverick-F35) [reply superagi/helper/webhook_manager.py:35]: renamed it

File: superagi/helper/webhook_manager.py (hunk 17 above)
Comment (Maverick-F35) [reply superagi/helper/webhook_manager.py:38]: removed

File: superagi/models/resource.py (hunk 18 above)
Comment (Maverick-F35) [reply superagi/models/resource.py:63]: renamed

File: superagi/models/project.py (hunk 21 above)
Comment (Maverick-F35) [reply superagi/models/project.py:60]: renamed

File: superagi/models/project.py (hunk 22 above)
Comment (Maverick-F35) [reply superagi/models/project.py:65]: renamed

File: superagi/models/toolkit.py (hunk 23 above)
Comment (Maverick-F35) [reply superagi/models/toolkit.py:160]: used get_toolkit_tools

File: superagi/models/web_hook_events.py (hunk 25 above)
Comment (Maverick-F35) [reply superagi/models/web_hook_events.py:7]: renamed

File: superagi/models/web_hooks.py [hunk 27]
```diff
@@ -0,0 +1,22 @@
+from sqlalchemy import Column, Integer, Text, String, Boolean, ForeignKey,JSON
+from sqlalchemy.orm import relationship
+from sqlalchemy.dialects.postgresql import JSONB
+from superagi.models.base_model import DBBaseModel
+from superagi.models.agent_execution import AgentExecution
+
+class WebHooks(DBBaseModel):
```
[the remark this answers is not in the document]
Comment (Maverick-F35) [reply superagi/models/web_hooks.py:7]: renamed

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py (hunk 20 above)
Comment (Maverick-F35) [reply migrations/versions/446884dcae58_add_api_key_and_web_hook.py:21]: renamed

File: migrations/versions/446884dcae58_add_api_key_and_web_hook.py (hunk 19 above)
Comment (Maverick-F35) [reply migrations/versions/446884dcae58_add_api_key_and_web_hook.py:24]: renamed

File: superagi/models/api_key.py [hunk 28]
```diff
@@ -0,0 +1,46 @@
...
+
+    @classmethod
+    def delete_by_id(cls, session,id: int):
+        db_api_key = session.query(ApiKey).filter(ApiKey.id == id).first()
+        db_api_key.is_expired = True
+        session.commit()
+        session.flush()
+
+    @classmethod
+    def edit_by_id(cls,session,id: int, name: str):
```
Comment (TransformerOptimus) [inline superagi/models/api_key.py:41]: rename it to update_api_key

Comment (Maverick-F35) [reply superagi/models/api_key.py]: renamed

**Review by Fluder-Paradyne — APPROVED** [review]

**Pull request merged**